src/TensorNetwork.jl

```diff
@@ -5,88 +5,92 @@ using OMEinsum
+struct TensorNetwork{D<:Domain,F<:NamedTuple}
     indices::Dict{Symbol,Vector{Int}}
     tensors::Vector{Tensor}
-    metadata::M
+    attr::F
 
-    function TensorNetwork{A}(tensors; metadata...) where {A}
+    function TensorNetwork{D}(tensors; attr...) where {D}
         indices = reduce(enumerate(tensors); init = Dict{Symbol,Vector{Int}}([])) do dict, (i, tensor)
             mergewith(vcat, dict, Dict([index => [i] for index in inds(tensor)]))
         end
 
-        # Check for inconsistent dimensions
+        # check dimension inconsistencies
         for (index, idxs) in indices
             allequal(Iterators.map(i -> size(tensors[i], index), idxs)) ||
-                throw(DimensionMismatch("Different sizes specified for index $index"))
+                throw(DimensionMismatch("Different dimensions specified for index $index"))
         end
 
-        M = Tenet.metadata(A)
-        metadata = M((; metadata...))
+        F = NamedTuple{inheritfieldnames(D),Tuple{inheritfieldtypes(D)...}}
+        attr = F((; attr...))
 
-        tn = new{A,M}(indices, tensors, metadata)
+        tn = new{D,F}(indices, tensors, attr)
+        checkdomain(tn)
 
-        checkansatz(tn)
         return tn
     end
 end
 
-TensorNetwork{A}(; metadata...) where {A<:Ansatz} = TensorNetwork{A}(Tensor[]; metadata...)
-
-# ansatz defaults to `Arbitrary`
-TensorNetwork(args...; kwargs...) = TensorNetwork{Arbitrary}(args...; kwargs...)
+TensorNetwork{D}(; attr...) where {D<:Domain} = TensorNetwork{D}(Tensor[]; attr...)
+TensorNetwork(args...; kwargs...) = TensorNetwork{Domain}(args...; kwargs...)
 
 # TODO maybe rename it as `convert` method?
-TensorNetwork{A}(tn::TensorNetwork{B}; metadata...) where {A,B} =
-    TensorNetwork{A}(tensors(tn); merge(tn.metadata, metadata)...)
-
-# TODO do sth to skip checkansatz? like @inbounds
-function checkansatz(tn::TensorNetwork{A}) where {A<:Ansatz}
-    for T in superansatzes(A)
-        checkmeta(T, tn) || throw(ErrorException("\"$T\" metadata is not valid"))
-    end
-end
+# TensorNetwork{A}(tn::TensorNetwork{B}; attr...) where {A,B} = TensorNetwork{A}(tensors(tn); merge(tn.attr, attr)...)
 
-checkmeta(::Type{<:Ansatz}, ::TensorNetwork) = true
-checkmeta(tn::TensorNetwork{T}) where {T<:Ansatz} = all(A -> checkmeta(A, tn), superansatzes(T))
+"""
+    domain(::TensorNetwork{Domain})
+    domain(::Type{<:TensorNetwork{Domain}})
 
-metadata(::Type{<:Ansatz}) = NamedTuple{(),Tuple{}}
-metadata(T::Type{<:Arbitrary}) = metadata(supertype(T))
+Return the `Domain` of a [`TensorNetwork`](@ref).
+"""
+domain(::TensorNetwork{D}) where {D} = D
+domain(::Type{<:TensorNetwork{D}}) where {D} = D
 
-Base.summary(io::IO, x::TensorNetwork) = print(io, "$(length(x))-tensors $(typeof(x))")
-Base.show(io::IO, tn::TensorNetwork) = print(io, "$(typeof(tn))(#tensors=$(length(tn)), #inds=$(length(tn.indices)))")
+# TODO do sth to skip `checkdomain`? like `@inbounds`
+checkdomain(tn::TensorNetwork{Domain}) = nothing
```

Comment:
We have to include the `NamedTuple` to be able to call `checkdomain` with a `TensorNetwork`:
```suggestion
checkdomain(tn::TensorNetwork{D, F}) where {D<:Domain, F<:NamedTuple} = nothing
```

Reply:
Nope, this must not work for subtypes of `Domain`. Only for `TensorNetwork{Domain}` itself. And `TensorNetwork{Domain,NamedTuple{(),Tuple{}}} <: TensorNetwork{Domain} <: TensorNetwork`, so no need to specify `F`.

Check out the `inheritcheckdomain` function. It recursively calls `checkdomain` on supertypes until it arrives `Domain` supertype.